lib/axon/loop.ex

```diff
@@ -1621,134 +1640,212 @@ defmodule Axon.Loop do
-
-                        {:continue, state} ->
-                          {:cont,
-                           {batch_fn, %{final_metrics_map | epoch => state.metrics},
-                            %State{
-                              state
-                              | epoch: epoch + 1,
-                                metrics: zero_metrics,
-                                iteration: 0,
-                                max_iteration: state.max_iteration
-                            }}}
-                      end
-                  end
-              end
+    * `:iterations` - max iterations to run each epoch. Must be non-negative
+      integer. Defaults to `-1` or no max iterations.
+
+    * `:jit_compile?` - whether or not to JIT compile initialization and step
+      functions. JIT compilation must be used for gradient computations. Defaults
+      to true.
+
+    * `:strict?` - whether or not to compile step functions strictly. If this flag
+      is set, the loop will raise on any cache miss during the training loop. Defaults
+      to true.
+
+    * `:debug` - run loop in debug mode to trace loop progress. Defaults to
+      false.
+
+    Additional options are forwarded to `Nx.Defn.jit` as JIT-options. If no JIT
+    options are set, the default options set with `Nx.Defn.default_options` are
+    used.
+  """
+  def run(%Loop{} = loop, data, init_state \\ %{}, opts \\ []) do
+    {max_epochs, opts} = Keyword.pop(opts, :epochs, 1)
+    debug? = Keyword.get(opts, :debug, false)
+
+    %Loop{output_transform: output_transform, handlers: handler_fns} = loop
+
+    state =
+      loop
+      |> stream(data, init_state, opts)
+      |> Stream.take(max_epochs)
+      |> Enum.reduce_while(:empty, fn
+        %{status: status} = last_epoch_state, current_state ->
+          event =
+            case status do
+              :halted -> :epoch_halted
+              :completed -> :epoch_completed
             end
-          )
-          |> case do
-            {final_metrics_map, state} -> {:halted, final_metrics_map, state}
-            {_batch_fn, final_metrics_map, state} -> {:completed, final_metrics_map, state}
+
+          case fire_event(event, handler_fns, last_epoch_state, debug?) do
+            {:continue, state} ->
+              {:cont, merge_states(state, current_state)}
+
+            {:halt_epoch, state} ->
+              {:cont, merge_states(state, current_state)}
+
+            {:halt_loop, state} ->
+              state = merge_states(state, current_state)
+              {:halt, %{state | status: :halted}}
           end
-      end
+      end)
 
-    {_, state} = fire_event(status, handler_fns, state, debug?)
-    state = %State{state | metrics: final_metrics}
+    {_, state} =
+      case state do
+        %{status: :completed} = state -> fire_event(:completed, handler_fns, state, debug?)
+        %{status: :halted} = state -> fire_event(:halted, handler_fns, state, debug?)
+      end
```

Comment:
We should convert this to `on_done` cause this state cannot be captured from a stream.